### PR TITLE
feat(react-client): migrate from recoil to zustand

### PR DESCRIPTION
**Pull request by aleks3fs**
There is an issue with running `@chainlit/react-client`, with latest React 19 and issue seems to be connected to usage of `recoil`.

In this PR I've 
* replaced `recoil` with `zustand`. I tried to combine logically connected parts of the state into stores.
* bumped React to v19

It'd be nice to have this in the latest version, but I haven't been able to test everything.

### CODE REVIEW

**Comment by aleks3fs** [comment]
If someone finds time to take this one over, since I might not have enough time, I'd appreciate it. Or if yours gets merged, I'm fine with both, as long as we get progress on compatibility with the latest version 😄

**Comment by aleks3fs** [comment]
Wonderful 🙏, I've sent you an invite for my fork, but do whatever is easier for you. I don't really care, as long as we get some progress on this 😄

**Comment by tegarmonster** [comment]
any update for this? :)

**Comment by FlawaCLV** [comment]
Any update? Any help needed?

**Comment by jbeckerdm** [comment]
@FlawaCLV @tegarmonster 
First and foremost we need a reviewer, because I'd like to know if the implementation is good or needs improvements.
Given how the chainlit team announced that they'll stop maintaining chainlit however, I feel like noone really wants to look into this at this time.

**Comment by FlawaCLV** [comment]
@jbeckerdm thanks for the update! I missed the news that they wanted to stop maintaining the project. I understand now why this PR is stalling.

**Comment by tegarmonster** [comment]
> @FlawaCLV @tegarmonster First and foremost we need a reviewer, because I'd like to know if the implementation is good or needs improvements. Given how the chainlit team announced that they'll stop maintaining chainlit however, I feel like noone really wants to look into this at this time.

ah I see, thanks for the information

**Comment by zomchak-code** [comment]
@jbeckerdm 
Could you share a link to this announcement?

> Given how the chainlit team announced that they'll stop maintaining chainlit however

**Comment by aleks3fs** [comment]
> @jbeckerdm Could you share a link to this announcement?
> 
> > Given how the chainlit team announced that they'll stop maintaining chainlit however

It's on the main page of the repo, 1st part of the readme 
https://github.com/Chainlit/chainlit/blob/main/backend/README.md?plain=1#L2

**Comment by jbeckerdm** [comment]
chatSettingsDefaultValue isn't properly populated yet, 
but perhaps we can define it in some other way (or well fix the issue).

If anyone wants to look into that, feel free!

**Comment by hayescode** [comment]
I am 1 of 2 users who received maintainer access. My strong suit is backend though and I'm not sure if the other maintainer has frontend experience. Can some of you review this and document any breaking changes/workarounds? This seems like a great PR but given the creator has left we're all in a tough spot. I don't want to further delay the huge backlog of improvements any further (just pushed 25% of outstanding PRs) but this is a big change that deserves more community feedback especially since this isn't my area.

**Comment by jbeckerdm** [comment]
@hayescode I'm essentially the new maintainer of this PR.
I'm trying to fix a very specific bug with the chat settings not loading (as in, the backend doesn't send the chat_settings websocket event to the frontend), which currently occurs in a test branch where I use this new code on our custom frontend/backend. It worked just fine with the newest chainlit react client and react 18, but now this hard to find bug occurs about half the time, and only in our CI test environment, not locally.
Since we didn't touch our backend in that branch, I suspect it's some edge case in this PR's code, so it's important to find it for this PR to get merged as well.

**Comment by jbeckerdm** [comment]
But I encourage anyone to test this PR in their own way and review the code! I'll make sure to work through the feedback.

**Comment by asvishnyakov** [comment]
@jbeckerdm Hi Jannis.

I'm the other maintainer, and while React isn’t the primary framework of choice used by me or my company, I do have frontend experience and a background in building frameworks.

I really appreciate your work, and in my opinion, the upgrade to React 19 looks good. However, switching from one state management library to another is a breaking change - especially since you’re modifying the shared `react-client` library. This library is used by developers who can't or don't want to use the full Chainlit frontend. Making such a change could create a cascade of compatibility issues in their projects, which I’d like to avoid.

Regarding performance: it’s great if switching to `zustand` brings improvements. I submitted a similar performance-focused [PR](https://github.com/Chainlit/chainlit/pull/1790) on the backend side, and it wasn’t accepted right away and for good reason. I think we need more measurable evidence before making changes like this, such as writing performance benchmarks or tests.

So, I’d like to propose the following plan:

1. Split out the React 19 upgrade into a separate PR. Or, if you don’t have the time, I can take care of it later.
2. Create an issue in the Chainlit repo describing the bug you're experiencing, including as many details as possible. If we’re able to reproduce the issue and determine that switching to `zustand` is the only viable fix, or if we gather clear performance data showing a benefit, we can revisit this PR and move forward with it.

Thanks again for your contribution!

**Comment by aleks3fs** [comment]
> @jbeckerdm Hi Jannis.
> 
> I'm the other maintainer, and while React isn’t the primary framework of choice used by me or my company, I do have frontend experience and a background in building frameworks.
> 
> I really appreciate your work, and in my opinion, the upgrade to React 19 looks good. However, switching from one state management library to another is a breaking change - especially since you’re modifying the shared `react-client` library. This library is used by developers who can't or don't want to use the full Chainlit frontend. Making such a change could create a cascade of compatibility issues in their projects, which I’d like to avoid.
> 
> Regarding performance: it’s great if switching to `zustand` brings improvements. I submitted a similar performance-focused [PR](https://github.com/Chainlit/chainlit/pull/1790) on the backend side, and it wasn’t accepted right away and for good reason. I think we need more measurable evidence before making changes like this, such as writing performance benchmarks or tests.
> 
> So, I’d like to propose the following plan:
> 
> 1. Split out the React 19 upgrade into a separate PR. Or, if you don’t have the time, I can take care of it later.
> 2. Create an issue in the Chainlit repo describing the bug you're experiencing, including as many details as possible. If we’re able to reproduce the issue and determine that switching to `zustand` is the only viable fix, or if we gather clear performance data showing a benefit, we can revisit this PR and move forward with it.
> 
> Thanks again for your contribution!

Hey, 

while performance boost is cool, the base idea of this PR was to enable upgrade to React 19. Latest version of `recoil` doesn't support react 19, and that is because `recoil` is deprecated. It was always an experimental library, so it was bound to happen.

It's really easy to dismiss performance improvement, if there isn't a bottleneck, but this one not it, at least not at its core.

Basically, if chainlit doesn't change from `recoil` to something else, then we cannot upgraade chainlit's react to 19, which means that any developer using chainlit's lib won't be able to do it either.

**Comment by asvishnyakov** [comment]
@aleks3fs Thank you for the explanation! I didn't see the deprecation message anywhere until I opened their repo. In that case, I think we don't have many options left, so I'll review this PR.

**Comment by jbeckerdm** [comment]
The aforementioned bug with chat_settings not loading has turned out to be a side effect of our custom frontend's connection logic, and is NOT related to this PR.
Which means that the PR is ready to be reviewed.

**Comment by neel004** [comment]
Hey @asvishnyakov, Great to have you on board as maintainer.
At the moment, the current Chainlit version makes it impossible for us to upgrade our React to 19. I noticed that this pull request might addresses this issue, I believe merging it would greatly benefit users who rely on React 19 compatibility.

May I kindly request your attention to reviewing and, if appropriate, merging this change at your earliest convenience? Your assistance would be greatly appreciated.
and @aleks3fs  @jbeckerdm  thanks for this pr

**Comment by neel004** [comment]
Just popping in to check if there’s any chance this PR could get a review soon. We’re kinda stuck on React 19 support without these changes, so merging this would really help out!

Let me know if there’s anything you need from my end, or if there’s something holding things up.

Thanks so much for keeping things moving — really appreciate all the work everyone’s putting in!

**Comment by hayescode** [comment]
CI needs to be updated to work with this

**Comment by asvishnyakov** [comment]
@hayescode I'm working on that [here](https://github.com/Chainlit/chainlit/pull/2327) and in few other PRs, let's wait until you return at Wednesday, I think it should be ready

File: libs/react-client/src/store/auth.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import { groupByDate } from 'src/utils/group';
+import { create } from 'zustand';
+
+import { IAuthConfig, IUser, ThreadHistory } from '..';
+
+interface AuthState {
+  authConfig?: IAuthConfig;
+  user?: IUser | null;
+
+  threadHistory?: ThreadHistory;
```
Comment (asvishnyakov) [inline libs/react-client/src/store/auth.ts:11]: While thread history is available only to authenticated users, I don't think it should be part of `AuthState`. It should be a separate state instead, as including it violates SRP.

File: libs/react-client/src/store/utils.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+export interface SetterFn<T> {
+  (
+    partial: T | Partial<T> | ((state: T) => T | Partial<T>),
+    replace?: false
+  ): void;
+  (state: T | ((state: T) => T), replace: true): void;
+}
+
+export const stateOrSetter = <T, K extends keyof T>(
```
Comment (asvishnyakov) [inline libs/react-client/src/store/utils.ts:9]: Why do we need this helper, if `set` already support this?
image

File: libs/react-client/src/store/chat.ts [hunk 3]
```diff
@@ -0,0 +1,121 @@
+import { ICallFn, ICommand, IMessageElement } from 'src/types';
+import { WavRecorder, WavStreamPlayer } from 'src/wavtools';
+import { create } from 'zustand';
+
+import { stateOrSetter } from './utils';
+
+// import { subscribeWithSelector } from 'zustand/middleware/subscribeWithSelector';
```
Comment (asvishnyakov) [inline libs/react-client/src/store/chat.ts:7]: Please remove a comment (did you forget something?)

File: libs/react-client/package.json [hunk 4]
```diff
@@ -1,10 +1,10 @@
 {
   "name": "@chainlit/react-client",
   "description": "Websocket client to connect to your chainlit app.",
-  "version": "0.2.4",
+  "version": "0.3.0",
```
Comment (asvishnyakov) [inline libs/react-client/package.json:4]: Please don't change version, we do it only at release time

File: libs/react-client/src/store/chat.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+    set({ chatSettingsInputs: [] });
+  },
+
+  resetChatSettingsValue: () => {
+    const chatSettingsValue = get().chatSettingsInputs.reduce(
+      (form: { [key: string]: any }, input: any) => (
+        (form[input.id] = input.initial), form
+      ),
+      {}
+    );
```
Comment (asvishnyakov) [inline libs/react-client/src/store/chat.ts:117]: Can we eliminate code duplication?

**Review by asvishnyakov — CHANGES_REQUESTED** [review]

**Comment by asvishnyakov** [comment]
So actually, after careful review, here are my two biggest concerns:

1. After merging this PR, we’ll be using React 19 in `libs/react-client` and React 18 in the frontend itself. This is highly risky in my opinion, especially since we’re not building or testing the frontend and libs against different React versions.

2. I made a few mistakes during merge and minor refactoring, and surprisingly the build didn’t fail - even though I removed a variable that was still in use, which led to a runtime error. Only the E2E tests caught it. I’d like to address this issue first.

@hayescode FYI

**Comment by mb-dbridge** [comment]
Hello! I can see this is still open with a 2.8.0 milestone while 2.8.1 has been released.
I'm wondering if there's any plan on re-prioritising this, or if there are any other plans?
Thanks for any update you could provide.

**Comment by goodslav** [comment]
Is this migration going to be finished soon? This is a very important migration for the future of the Chainlit FE given the fact that Recoil most likely is obsolete as a project and will never support React 19 so this is actually also blocking the upgrade of the project to a higher version of React.

**Comment by mayurw-pmam** [comment]
Any update on this?

File: libs/copilot/src/chat/index.tsx [hunk 6]
```diff
@@ -1,20 +1,15 @@
 import { useEffect, useRef } from 'react';
-import { useRecoilValue, useSetRecoilState } from 'recoil';
+import { useRecoilValue } from 'recoil';
```
Comment (nnmer) [inline libs/copilot/src/chat/index.tsx:2]: here is probably forgotten 'recoil' use

**Comment by jbeckerdm** [comment]
I won't find time to continue work on this, if anyone wants to take it over or do another take on the zustand migration, let me know. I'll leave the sources up, but feel free to open another PR if that's easier.

**Comment by dokterbob** [comment]
Closing as stale.
